File: 02_25/PLP/tai_5/main.py
'''
TAI-5 -> Paradigma Orientado a Objeto
Sistema de Produtos
'''

# classe base
class Produto:
    def __init__(self, nome):
        self.nome = nome
        self._preco = -1.0

    # método de acesso
    def get_preco(self):
        return self._preco
    
    def set_preco(self, novo_preco):
        if novo_preco >= 0:
            self._preco = novo_preco
        else:
            print(f'ERRO: O preço do produto {self.nome} não pode ser negativo.')

    def exibir_Detalhes(self):
        if self._preco > 0:
            print(f'Produto : {self.nome}   |   Preço : R${self._preco:.2f}')
        elif self._preco == 0:
            print(f'Produto : {self.nome}   |   Preço : Gratuito')
        else:
            print(f'Produto : {self.nome}   |   Preço : Não Informado.')

class Eletrônico(Produto):
    def __init__(self, nome, voltagem):
        super().__init__(nome)
        self.voltagem = voltagem

    def exibir_Detalhes(self):
        print(f'Produto : {self.nome}   |   Preço : R${self._preco:.2f}   |   Voltagem : {self.voltagem}w')


if __name__ == '__main__':
    p1 = Produto('Caderno')
    p1.set_preco(18.50)

    p2 = Produto('Tênis')
    p2.set_preco(0.0)

    p3 = Produto('Par de Meias')
    p3.set_preco(-10)

    e1 = Eletrônico('Liquidificador', 110)
    e1.set_preco(229.99)
    
    e2 = Eletrônico('Micro-ondas', 220)
    e2.set_preco(149.49)

    p1.exibir_Detalhes()
    p2.exibir_Detalhes()
    p3.exibir_Detalhes()
    e1.exibir_Detalhes()
    e2.exibir_Detalhes()

File: 02_25/PLP/tai_5/Main.java
/*
 * TAI-5 -> Paradigma Orientado a Objeto
 * Sistema de Produtos
 */

// Classe base
class Produto {
    // Atributos privados (Encapsulamento)
    private String nome;
    private double preco;

    // Construtor
    public Produto(String nome) {
        this.nome = nome;
        this.preco = -1.0;
    }

    // Métodos de acesso (Getters e Setters)
    public String getNome() {
        return this.nome;
    }

    public double getPreco() {
        return this.preco;
    }

    public void setPreco(double novoPreco) {
        if (novoPreco >= 0) {
            this.preco = novoPreco;
        } else {
            System.out.printf("ERRO: O preço do produto %s não pode ser negativo.%n", this.nome);
        }
    }

    // Método para exibir detalhes
    public void exibirDetalhes() {
        if (this.preco > 0) {
            System.out.printf("Produto : %s   |   Preço : R$%.2f%n", this.nome, this.preco);
        } else if (this.preco == 0) {
            System.out.printf("Produto : %s   |   Preço : Gratuito%n", this.nome);
        } else {
            System.out.printf("Produto : %s   |   Preço : Não Informado.%n", this.nome);
        }
    }
}

// Classe derivada (Herança)
class Eletronico extends Produto {
    private int voltagem;

    public Eletronico(String nome, int voltagem) {
        super(nome); // Chama o construtor da classe pai
        this.voltagem = voltagem;
    }

    public void exibirDetalhes() {
        // Como os atributos são privados na classe pai, usamos os getters (getNome e getPreco)
        System.out.printf("Produto : %s   |   Preço : R$%.2f   |   Voltagem : %dw%n", 
                          this.getNome(), this.getPreco(), this.voltagem);
    }
}

// Classe principal para execução (Main)
public class Main {
    public static void main(String[] args) {
        Produto p1 = new Produto("Caderno");
        p1.setPreco(18.50);

        Produto p2 = new Produto("Tênis");
        p2.setPreco(0.0);

        Produto p3 = new Produto("Par de Meias");
        p3.setPreco(-10);

        Eletronico e1 = new Eletronico("Liquidificador", 110);
        e1.setPreco(229.99);
        
        Eletronico e2 = new Eletronico("Micro-ondas", 220);
        e2.setPreco(149.49);

        System.out.println();
        p1.exibirDetalhes();
        p2.exibirDetalhes();
        p3.exibirDetalhes();
        e1.exibirDetalhes();
        e2.exibirDetalhes();
    }
}